File: cmd/cs.rb
#!/usr/bin/env ruby

def usage()
  STDERR.puts("usage: #{File.basename(__FILE__)} o n [m] [p]")
  exit(1)
end

if ARGV.length < 2 or ARGV.length > 4
  usage()
end

o = ARGV[0].to_i()
if o < 0
  usage()
end

n = ARGV[1].to_i()
if n <= 0
  usage()
end

m = 1
if ARGV.length >= 3
  m = ARGV[2].to_i()
  if m <= 0 or m > n
    usage()
  end
end

p = m
if ARGV.length == 4
  p = ARGV[3].to_i()
  if p < m or p > n
    usage()
  end
end

i = 0
while i < n
  unless File.exists?("#{i}.txt.tfidf")
  	exit(3)
  end
  i += 1
end

csns = Array.new(m)
pp = 0
pq = 0
pc = 0
ps = 0

i = 0
j = (n.to_f() / p.to_f()).ceil()
while i < n
  if pc < m
    k = i + j
    b = [k, n].min
    c = "cs#{o} #{n} #{i} #{b}"
    d = c.gsub(/\s+/, '_')
    csn = Process.fork do
      exec("#{c} > #{d}.out 2> #{d}.err")
    end
    csns[pq % m] = csn
    pq += 1
    STDERR.puts(c)
    pc += 1
    i = k
  end
  if pc == m
    csn = csns[pp % m]
    pp += 1
    Process.wait(csn)
    s = $?.exitstatus()
    if s != 0
      ps = s
    end
    pc -= 1
  end
end

while pc > 0
  csn = csns[pp % m]
  pp += 1
  Process.wait(csn)
  s = $?.exitstatus()
  if s != 0
    ps = s
  end
  pc -= 1
end

exit(ps)

File: cmd/cs.go
package main

import (
	"fmt"
	"math"
	"os"
	"os/exec"
	"path"
	"strconv"
	"strings"
)

func usage() {
	fmt.Fprintf(os.Stderr, "usage: %s o n m p\n", path.Base(os.Args[0]))
	os.Exit(1)
}

func main() {
	if len(os.Args) != 5 {
		usage()
	}

	o, err := strconv.ParseInt(os.Args[1], 10, 32)
	if err != nil || o < 0 {
		usage()
	}

	n, err := strconv.ParseInt(os.Args[2], 10, 32)
	if err != nil || n <= 0 {
		usage()
	}

	m, err := strconv.ParseInt(os.Args[3], 10, 32)
	if err != nil || m <= 0 || m > n {
		usage()
	}

	p, err := strconv.ParseInt(os.Args[4], 10, 32)
	if err != nil || p < m || p > n {
		usage()
	}

	for i := int64(0); i < n; i++ {
		p := fmt.Sprintf("%d.txt.tfidf", i)
		if _, err := os.Stat(p); os.IsNotExist(err) {
			fmt.Fprintf(os.Stderr, "%s\n", err)
			os.Exit(3)
		}
	}

	csns := make([]*exec.Cmd, m)
	pc := 0
	pp := 0
	pq := 0
	ps := 0

	i := int64(0)
	j := int64(math.Ceil(float64(float32(n) / float32(p))))
	for i < n {
		if int64(pc) < m {
			k := i + j
			a := i
			b := int64(math.Min(float64(k), float64(n)))
			c := fmt.Sprintf("cs%d %d %d %d", o, n, a, b)
			d := strings.Replace(c, " ", "_", -1)
			csn := exec.Command("/bin/sh", "-c", c)
			fn := fmt.Sprintf("%s.out", d)
			of, err := os.OpenFile(fn, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0666)
			if err != nil {
				fmt.Fprintf(os.Stderr, "%s\n", err)
				os.Exit(4)
			}
			csn.Stdout = of
			fn = fmt.Sprintf("%s.err", d)
			ef, err := os.OpenFile(fn, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0666)
			if err != nil {
				fmt.Fprintf(os.Stderr, "%s\n", err)
				os.Exit(4)
			}
			csn.Stderr = ef
			err = csn.Start()
			if err == nil {
				of.Close()
				ef.Close()
				csns[pq%cap(csns)] = csn
				pq++
				fmt.Fprintln(os.Stderr, c)
				pc++
				i = k
			}
		}
		if int64(pc) == m {
			csn := csns[pp%cap(csns)]
			pp++
			err := csn.Wait()
			if err != nil {
				ps = 5
			}
			pc--
		}
	}

	for int64(pc) > 0 {
		csn := csns[pp%cap(csns)]
		pp++
		err := csn.Wait()
		if err != nil {
			ps = 5
		}
		pc--
	}

	os.Exit(ps)
}
